### ISSUE TITLE
Allowing per-request headers in C# client

### ISSUE THREAD

**Issue by Pzixel** [RicoSuter/NSwag#2417]
My use-cases required setting some headers per request (e.g. Bearer token, RequestId and Referrer header, etc..).

Unfortunately this forces me to re-create a HttpClient per-request so I can't create it once and for all.

I'd like to have a generation option like `GenerateHttpMessageAlteration` which would add `Action<HttpRequestMessage>` to every message to alterate all required properties (e.g. headers).

I imagine that I could call them like `generatedClient.SomeController_SomeMethodparameter, MyHelper.SetDefaultHeaders(requestId, referrer))`.

I can create a PR if it fits your vision on the project. But currently I can't inject anything and I have to create a clients manually everywhere. I'd like to create them only once via DI and only alterate some per-request features when needed.

---

This is how it looks like now:

```cs
ctor() {}

Task HandleRequestAsync(string someParam, string principalToken, stirng requestId) 
{
    var client = new MyClient(GetHttpClient(Constants.MyClientApiUrl, principalToken, requestId, nameof(MyHandler)));
    return client.DoSomething(someParam);
}

private static HttpClient GetHttpClient(string apiUrl, string principalToken, string requestId, string refferer)
{
    var httpClient = new HttpClient
    {
        BaseAddress = new Uri(apiUrl),
        DefaultRequestHeaders =
        {
            Authorization = new AuthenticationHeaderValue("Bearer", principalToken),
        }
    };

    httpClient.DefaultRequestHeaders.TryAddWithoutValidation("RequestId", requestId);
    httpClient.DefaultRequestHeaders.TryAddWithoutValidation("Referer", refferer);
    return httpClient;
}
```

This is how it could be like:

```cs
ctor(IMyClient myClient) 
{
    this.myClient = myClient;
}

Task HandleRequestAsync(string someParam, string principalToken, stirng requestId) 
{
    return myClient.DoSomething(someParam, AlterateHeaders(principalToken, requestId, nameof(MyHandler)));
}

private static Action<HttpRequestMessage> AlterateHeaders(string principalToken, string requestId, string refferer) =>
    request => {
        request.Headers.Add("RequestId", requestId);
        request.Headers.Add("Referer", refferer);
        request.Headers.Add("Authorization", GetAuthHeader(principalToken));
    };
```

In the latter case we could reuse the single `Client` accross the whole application, we only change some requests as required per our logic.

**Comment by Pzixel** [comment]
Hey! Any ideas on that? Is it "won't do", "we will take this into account" or anything?

**Comment by RicoSuter** [comment]
https://github.com/RicoSuter/NSwag/blob/master/src/NSwag.Commands/Commands/CodeGeneration/OpenApiToCSharpClientCommand.cs#L120

**Comment by Pzixel** [comment]
Oh, sorry, I didn't find that in docs. Excuse me for a dumb question.

**Comment by rvarna** [comment]
@Pzixel and @RicoSuter I apologize for digging up an old thread, but even with the UseHttpRequestMessageCreationMethod, it only allows creating the RequestMessage and associated headers from the base class. If we are using a single generate client across the application (which is injected as a singleton via Dependency Injection) and we need to add per request headers, I don't understand it can be done in a thread safe way. Even if the base class takes another object as dependency, that would still be singleton and cannot have "scoped" dependency in .Net Core. I am perhaps missing something, but can you elaborate a little more (I can understand if the same headers need to be added to every request, but if the headers need to be generated based on some non-singleton data, how do we do that)?

Here is an example:
Service A gets called with some data from a public client and then, based on the request, calls a downstream service B using the NSwag client. However, the headers for this request are generated from the request to service A and will differ for each request.

**Comment by jmevel** [comment]
> [@Pzixel](https://github.com/Pzixel) and [@RicoSuter](https://github.com/RicoSuter) I apologize for digging up an old thread, but even with the UseHttpRequestMessageCreationMethod, it only allows creating the RequestMessage and associated headers from the base class. If we are using a single generate client across the application (which is injected as a singleton via Dependency Injection) and we need to add per request headers, I don't understand it can be done in a thread safe way. Even if the base class takes another object as dependency, that would still be singleton and cannot have "scoped" dependency in .Net Core. I am perhaps missing something, but can you elaborate a little more (I can understand if the same headers need to be added to every request, but if the headers need to be generated based on some non-singleton data, how do we do that)?
> 
> Here is an example: Service A gets called with some data from a public client and then, based on the request, calls a downstream service B using the NSwag client. However, the headers for this request are generated from the request to service A and will differ for each request.

That's exactly my use case and I'm also confused how to do that.

How did you implement that at the end?

**Comment by jmevel** [comment]
For anyone passing by here I found the solution.

Here is my code for dynamically forwarding a Bearer token only for specifics API calls (`IFooApiClient` is a client generated by NSwag if that matters) 

```csharp
public static IServiceCollection AddFooApiClient(this IServiceCollection services, IConfiguration configuration)
{
    return services.AddScopedHttpClientDependentService<IFooApiClient>(
        newService: static (baseUrl, httpClient) => new FooApiClient(baseUrl, httpClient),
        configuration: configuration,
        configurationSectionName: "ApiServices:Foo"
    );
}

private delegate TService InstanciateService<out TService>(string baseUrl, HttpClient httpClient);

private static IServiceCollection AddScopedHttpClientDependentService<TService>(
    this IServiceCollection services,
    InstanciateService<TService> newService,
    IConfiguration configuration,
    string configurationSectionName
)
    where TService : class
{
    //derives the httpClientName from section name to avoid any conflict on httpClientName
    string httpClientName = $"{configurationSectionName}HttpClient";

    var options = configuration.GetRequiredSection(configurationSectionName).Get<ApiConfiguration>() ?? throw new Exception("Unable to bind configuration");

    var httpClientBuilder = services
        .AddHttpClient(httpClientName)
        .AddHttpMessageHandler(sp => new BearerTokenForwardingMessageHandler(sp.GetRequiredService<IHttpContextAccessor>()));

    services.AddScoped(provider =>
    {
        var httpClientFactory = provider.GetRequiredService<IHttpClientFactory>();
        var httpClient = httpClientFactory.CreateClient(httpClientName);
        return newService(options.EndpointUrl.OriginalString, httpClient);
    });

    return services;
}

public sealed class BearerTokenForwardingMessageHandler(IHttpContextAccessor httpContextAccessor) : DelegatingHandler
{
    protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
    {
        if (TryGetAccessToken(httpContextAccessor, out var token))
        {
            request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);
        }
        return await base.SendAsync(request, cancellationToken);
    }

    private static bool TryGetAccessToken(IHttpContextAccessor httpContextAccessor, out string? accessToken)
    {
        accessToken = null;
        if (!httpContextAccessor.HttpContext!.Request.Headers.TryGetValue("Authorization", out StringValues authorizationHeader))
        {
            return false;
        }
        var bearerHeader = authorizationHeader.FirstOrDefault(h => h is not null && h.StartsWith("Bearer", StringComparison.OrdinalIgnoreCase));
        if (bearerHeader is null)
        {
            return false;
        }

        accessToken = bearerHeader["Bearer ".Length..];
        return true;
    }
}
```
